### ISSUE TITLE
--prune-degenerate=always can result in commit different from original

### ISSUE THREAD

**Issue by rndbit** [newren/git-filter-repo#313]
Hello.

When a merge is made with `-s ours` as in `git merge -s ours other-branch` and then doing git-filter-repo with `--prune-degenerate=always` the merge commit gets pruned (not so bad), but the child commit of the pruned commit ends up using the 2nd parent of the pruned commit instead of the 1st parent (1st parent is identical to the pruned one) and it breaks branch content by bringing changes from other-branch, which is what was being avoided by merging with `-s ours`.

Here is a test script:

```sh
#!/bin/bash

set -e
set -u
set pipefail
set -x

REPO=test-prune-degenerate.git

function gitgraph() {
  git log --graph --decorate -M --parents --name-status --oneline --format="%C(auto)h=%h t=%t p=[%p] %d %s" --all
}

rm -rf "${REPO}"
mkdir -pv "${REPO}"
git init "${REPO}"
cd "${REPO}"

echo 1 > common-file1
echo 1 > common-file2
git add -A
git commit -m 'A start'

git branch maint-rel-1
git branch maint-rel-2

# Now make a change on a maintenance v2 branch, but this change breaks v1 so must not make it to maint-rel-1
git checkout maint-rel-2
git rm common-file1
git commit -m 'B delete on maint v2 breaks v1

this change breaks v1 and must not make it across to maint-rel-1'

git checkout maint-rel-1
git merge -s ours maint-rel-2 -m 'C noop merge from maint-rel-2

nmark as merged from maint-rel-2 to skip brining breaking change,
this allows to continue merging from maint-rel-2 to backport other
changes that will follow after the breaking commit

after running git-filter-repo --prune-degenerate=always this change
ends up in maint-rel-1 anyway!!!!!!
'

# So far is enough for the fault demonstration, but add one more commit to make it clearer
echo 2 >> common-file2
git add -A
git commit -m 'D more fixes on v1 maintenance'

# Done with setup, now do filtering

gitgraph # show the initial graph
[ -e common-file1 ] || { echo "FAIL 1"; exit 1; } # Ensure test passes initially

# This --prune-degenerate=always breaks maint-rel-1 by including changes from maint-rel-2
git filter-repo --force --prune-degenerate=always
git checkout maint-rel-1

gitgraph # show the graph AFTER filtering
[ -e common-file1 ] || { echo "FAIL 2"; exit 2; } # This test fails now

echo "ALL done OK" # Does not run, exits earlier
```

Before filtering the graph is:

```txt
A---C---D  <== maint-rel-1
 \ /
  B <== maint-rel-2

# content of A and C is identical due to merge with -s ours
```

After filtering the graph is:
```txt
A   D  <== maint-rel-1
 \ /
  B <== maint-rel-2

# but D' != D, it contains changes from B, the changes that were being avoided.
```

Good outcomes would be, either no pruning of commit C, or graft D onto A (1st parent of pruned C) instead of grafting D onto B. So this would be OK:

```txt
A------D  <== maint-rel-1
 \ 
  B <== maint-rel-2
```

I had a look at the code and at the output from `git fast-export`. Unfortunately export does not contain differences against 2+ parents, only differences against 1st parent, so I don't see how to figure this out. Hopefully someone might have a clever idea though.

I understand that this problem can be avoided with `--no-ff` passed to get-filter-repo. This is not ideal. There are several trees that after pruning empty simple commits end up with merges looking like --no-ff and it would be nice to tidy those up too if this could be done without breaking occasional tricky merges like shown here.

IMO ideally such tricky merges would not be pruned. The fake-merge as in the example served a purpose of preventing breaking change from being merged for real.

**Comment by newren** [comment]
I think this is fixed with 912dbca53e91 (filter-repo: fix corner case involving bad pruning of evil degenerate merge, 2024-10-18); could you verify?

**Comment by rndbit** [comment]
I confirm that [912dbca](https://github.com/newren/git-filter-repo/commit/912dbca53e91392db65ca557529393e8e97b846d) does fix this issue at least according to the test included in the original issue description. Thank you!
